torchtext/experimental/transforms.py

```diff
@@ -340,3 +341,147 @@ def forward(self, tokens: List[str]) -> Tensor:
+                [6, 0, 1, 2, 3], [6, 0, 1, 4, 4]
+            ])
+        >>> mask_transform = MaskTransform(
+                vocab_len = 7,
+                mask_idx = 4,
+                bos_idx = 6,
+                pad_idx = 5,
+                mask_bos = False,
+                mask_prob = 0.15
+            )
+        >>> masked_tokens, target_tokens, mask = mask_transform(sample_token_ids)
+    """
+
+    # maks_mask_prob is prob. of replacing a token with [MASK] (ex. 80%)
+    mask_mask_prob = 0.8
+
+    # rand_mask_thresh is prob. of replacing a token with a random token. (ex.10%)
+    rand_mask_prob = 0.1
+
+    def __init__(
+        self,
+        vocab_len: int,
+        mask_idx: int,
+        bos_idx: int,
+        pad_idx: int,
+        mask_bos: bool = False,
+        mask_prob: float = 0.15,
+    ):
+        super().__init__()
+        self.vocab_len: int = vocab_len
+        self.mask_idx: int = mask_idx
+        self.bos_idx: int = bos_idx
+        self.pad_idx: int = pad_idx
+        self.mask_prob: float = mask_prob
+        self.mask_bos: bool = mask_bos
+
+    def forward(self, tokens: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """
+        Applies mask to input tokens.
+
+        Inputs:
+            tokens: Tensor with token ids of shape (batch_size x seq_len)
+
+        Outputs:
+            masked_tokens: Tensor of tokens after masking has been applied
+            target_tokens: Tensor of token values selected for masking
+            mask: Tensor with same shape as input tokens (batch_size x seq_len)
+                with masked tokens represented by a 1 and everything else as 0.
+        """
+        # tokens, mask, mask_mask, rand_mask: (T, C)
+        mask, mask_mask, rand_mask = self._generate_mask(tokens)
+
+        # a. generate the masked input tokens
+        # (1) the [MASK] token 80% of the time
+        masked_tokens = self._mask_input(tokens, mask_mask, self.mask_idx)
+        # (2) a random token 10% of the time
+        masked_tokens = self._mask_input(
+            masked_tokens,
+            rand_mask,
+            torch.randint_like(tokens, high=self.vocab_len),
+        )
+
+        # b. generate the target prediction
+        target_tokens = torch.masked_select(tokens, mask.bool())
+
+        # masked_tokens: (T, C), target_tokens: (T x C x mask_prob, ), mask
+        return masked_tokens, target_tokens, mask
+
+    def _random_masking(self, tokens: torch.tensor, mask_prob: float) -> torch.Tensor:
+        """
+        Function to mask tokens randomly.
+
+        Inputs:
+            1) tokens: Tensor with token ids of shape (batch_size x seq_len)
+            2) mask_prob: Probability of masking a particular token
+
+        Outputs:
+            mask: Tensor with same shape as input tokens (batch_size x seq_len)
+                with masked tokens represented by a 1 and everything else as 0.
+        """
+        batch_size, seq_len = tokens.size()
```

Comment:
nit: Looks like tokens are used only to get the sizes. Should we pass the size directly instead?

Reply:
@parmeet I think just a few lines below it is invoked to convert the mask from a numpy array to a tensor (as shown below). I suppose we could change it by explicitly stating whether we want the tensor to be allocated on a cpu or cuda device, though I'm not sure if it must match the device on which `tokens` is allocated (which as far as I've checked isn't explicitly defined in the script)

`mask = torch.from_numpy(mask).to(tokens.device)`

Reply:
To follow up, this issue may be resolved if we refactor to remove the numpy dependency entirely (as suggested by @mthrok in later comments)